### DISCUSSION TITLE
Subscribe user with token

### DISCUSSION THREAD

**Discussion by ArensMyzyri** [dunglas/mercure#604] [Q&A]
Hi. 
I have read through documentation and hitting a wall on understanding how to authenticate a subscriber to subscribe to a private topic. 

Here my PHP  part: 
```
 $update = new Update(
                $message->getTopic()->getName(),
                $message->getTopic()->getData(),
                true
 );
            
 $this->publisher->publish($update);
```

The problem I have and  that I can't understand is how can I add a token to this topic? 
How to make it so the front end can subscribe if has this token. 
I see no option above to create/update a topic with a dynamic token which the front end client can use together with the topic to subscribe. 
Ex: 
https://localhost/.well-known/mercure?topic=myTopicName
Authorization: "Bearer SomeToken"

Please let me know if I'm missing something.

**Comment by dunglas** [comment]
Can you share the topic you use (the string) and as well as the JWT?

**Reply by ArensMyzyri** [reply]
Hi there. 

Subscribing using a token: 
![image](https://user-images.githubusercontent.com/22968942/148276707-6c448e5d-218a-4a71-910f-4151d9c36d62.png)

So here I create/update the topic: 
![image](https://user-images.githubusercontent.com/22968942/148276531-6ce211fd-ed36-486a-96fd-f3e5131d7afd.png)


If it has private: true, will never be received by the subscriber.

the token/key is the same for both publisher and subscriber. 

JWT:
eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJtZXJjdXJlIjp7InB1Ymxpc2giOlsiYWRkUHJvZHVjdFRpdGxlIl19fQ.GB8yj7zhtYFw9-z94YGfQEGOqDdKMCqCb3E_7eYzTSU

**Reply by ArensMyzyri** [reply]
Oh made it work now :D
So, writing here, if anyone else might run into same trouble of understanding same as did I. 

When you generate a token to send to the client in the front end, do this: 
```
$token = JWT::encode([
                'mercure' => [
                    'subscribe' => ['myTopic'],
                ],
            ], 'mySecret');
            
             return $token
```
JWT::encode is from Firebase\JWT;

Then, this token need to be used in the Authorization with Bearer, or in the cookie.

PS. if you do: 
  ```
JWT::encode([
                'mercure' => [
                    'publish' => ['myTopic'],
                ],
            ], 'mySecret');
```
then this token can be used to publish a topic. 
Cheers :)
